### PR TITLE
Issue 151: Add a file-watcher API

### PR DESCRIPTION
**Pull request by robstryker**
Tests and cleanup of model

Signed-off-by: Rob Stryker <rob@oxbeef.net>

Some failing tests, weird osgi stuff

Signed-off-by: Rob Stryker <rob@oxbeef.net>

Unit tests

Signed-off-by: Rob Stryker <rob@oxbeef.net>

### CODE REVIEW

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+		} catch (IOException e) {
+			log(e);
+		}
+		if( watchService == null )
+			throw new IllegalStateException();
+		serviceThread = new Thread(this);
+		serviceThread.start();
+	}
+	
+	public void stop() {
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: missing @Override

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+
+	private synchronized void setClosing(boolean b) {
+		this.closing = b;
+	}
+	
+	private synchronized boolean isClosing() {
+		return this.closing;
+	}
+	
+	public void start() throws IllegalStateException {
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: missing @Override

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+	
+	private void clearModel() {
+		for( WatchKey key : subscriptions.values()) {
+			key.cancel();
+		}
+		subscriptions.clear();
+	}
+	
+	
+	public synchronized void registerListener(Path path,
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: missing @Override

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+				subscriptions.put(working, k);
+				if( existing != null )
+					existing.cancel();
+			}
+		} catch (IOException e) {
+			log(e);
+		}
+	}
+
+	public synchronized void deregisterListener(Path path, IFileWatcherEventListener listener) {
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: missing @Override

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+		if( list != null ) {
+			Iterator<RegistrationRequest> rit = list.iterator();
+			RegistrationRequest r = null;
+			while(rit.hasNext()) {
+				r = rit.next();
+				if( r.getListener() == listener ) {
+					rit.remove();
+				}
+			}
+			if( list.size() == 0 ) {
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: ```list.isEmpty()```

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+	}
+
+
+	private boolean pathShouldBeSubscribed(Path path) {
+		// A recursive request to me or my parent exists, so I'm still needed
+		if( pathOrParentHasRecursiveRequest(path))
+			return true;
+		
+		// A request still exists for this exact path, so I'm still needed
+		if( requests.get(path) != null && requests.get(path).size() > 0 )
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: ```!requests.get(path).isEmpty()```

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+						Path context = ((Path)key.watchable()).resolve((Path)event.context());
+						subscribeToChanges(event, context);
+						fireEvents(key, event);
+					}
+				}
+			}
+		} catch (InterruptedException | ClosedWatchServiceException e) {
+			if( isClosing() == false ) {
+				// unexpected error
+				e.printStackTrace();
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: replace with real logging

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+					Path eventContext = (Path)event.context();
+					if( eventContext != null ) {
+						Path context = ((Path)key.watchable()).resolve((Path)event.context());
+						subscribeToChanges(event, context);
+						fireEvents(key, event);
+					}
+				}
+			}
+		} catch (InterruptedException | ClosedWatchServiceException e) {
+			if( isClosing() == false ) {
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: ```if(!isClosing())```

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+					subscribeSinglePath(eventContext);
+				}
+			}
+		}
+		// ignore modify event kind
+		
+	}
+
+	private boolean requestMatchesExact(Path path) {
+		return requests.get(path) != null && requests.get(path).size() > 0;
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: ```!requests.get(path).isEmpty();```

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+	 * both the parameter and all of its parent directories
+	 */
+	private Path[] getPathAndAncestors(Path p) {
+		ArrayList<Path> ret = new ArrayList<>();
+		Path working = p;
+		while( working != null ) {
+			ret.add(working);
+			working = working.getParent();
+		}
+		return (Path[]) ret.toArray(new Path[ret.size()]);
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: unnecessary cast ```(Path[])```

File: framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/filewatcher/IFileWatcherService.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.nio.file.Path;
+
+public interface IFileWatcherService {
+
+	public void start() throws IllegalStateException;
+	
+	public void stop();
+	
+	public void registerListener(Path path,
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/filewatcher/IFileWatcherService.java]: I'd rather call this method ```addFileWatcherEventListener()``` to be consistent with the rest of the API.
ex. In ServerModel
```
public void addServerModelListener(IServerModelListener l) {
```

Having a consistent naming scheme helps a lot to get accustomed to an API.

File: framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/filewatcher/IFileWatcherService.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+public interface IFileWatcherService {
+
+	public void start() throws IllegalStateException;
+	
+	public void stop();
+	
+	public void registerListener(Path path, 
+			IFileWatcherEventListener listener, boolean recursive);
+	
+	public void deregisterListener(Path path, IFileWatcherEventListener listener);
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/filewatcher/IFileWatcherService.java]: I'd rather call this method ```removeFileWatcherEventListener()``` to be consistent with the rest of the API.
ex. In ServerModel
```
public void removeServerModelListener(IServerModelListener l) {
```

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 13]
```diff
@@ -0,0 +1,584 @@
...
+					}
+				}
+			}
+		} catch (InterruptedException | ClosedWatchServiceException e) {
+			if( isClosing() == false ) {
+				// unexpected error
+				e.printStackTrace();
+			}
+		}
+		setClosing(false);
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: why are we overriding a possible ```closing == true```here?

ex. the thread is interrupted (it is being shut down), then we're setting closing to false and continue? What am I missing?

File: distribution/integration-tests/src/test/java/org/jboss/tools/rsp/itests/wildfly/WildFlyPublishingTest.java [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+        while(tries > 0) {
+            tries--;
+            String s = client.getStateString();
+            System.out.println("State String: " + s);
+            if (serverState.equals(s)) {
+            	client.clearState();
+                return;
+            }
+            Thread.sleep(1000);
+        }
```
Comment (adietish) [inline distribution/integration-tests/src/test/java/org/jboss/tools/rsp/itests/wildfly/WildFlyPublishingTest.java]: l145-154 could be replaced by a much simpler timely limited poll from a ```ArrayBlockingQueue``` (that gets filled up by the listener) in the TestClient:
I did exactly that in ```PollThreadTest```:
https://github.com/redhat-developer/rsp-server/blob/master/framework/tests/org.jboss.tools.rsp.server.spi.test/src/main/java/org/jboss/tools/rsp/server/spi/model/polling/PollThreadTest.java#L336

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 15]
```diff
@@ -0,0 +1,584 @@
...
+		}
+		return false;
+	}
+
+	@Override
+	public void run() {
+		WatchKey key;
+		try {
+			while (watchService != null && (key = watchService.take()) != null) {
+				// Get the list and reset immedietely
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: misspelled ```immedietely```.
Can you please explain in the comment why you reset it (I imagine that this somehow marks the event as consumed?)?

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 16]
```diff
@@ -0,0 +1,584 @@
...
+				List<WatchEvent<?>> events = key.pollEvents();
+				key.reset();
+				
+				for (WatchEvent<?> event : events) {
+					Path eventContext = (Path)event.context();
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: don't we have to check the type of the event context (we're directly casting to ```Path``` without checking)

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java [hunk 17]
```diff
@@ -0,0 +1,584 @@
...
+	
+	public void start() throws IllegalStateException {
+		try {
+			watchService = FileSystems.getDefault().newWatchService();
+		} catch (IOException e) {
+			log(e);
+		}
+		if( watchService == null )
+			throw new IllegalStateException();
+		serviceThread = new Thread(this);
```
Comment (adietish) [inline framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: I'd favor a more modern pattern where you submit ```this``` to an ExecutorService. The benefit is that the service can be forcefully shut down. 

See here in AbstractPoller: https://github.com/redhat-developer/rsp-server/blob/master/framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/model/polling/AbstractPoller.java#L40

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java (hunk 1 above)
Comment (adietish) [reply framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: shouldn't ```stop()``` be synchronized (or at least ```clearModel()``` and closing the watchService, too)?

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java (hunk 2 above)
Comment (adietish) [reply framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: shouldn't ```start()``` be synchronized?

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+import java.util.HashMap;
+import java.util.List;
+
+import org.jboss.tools.rsp.server.spi.filewatcher.IFileWatcherEventListener;
+import org.junit.Test;
+
+public class FilewatcherModelTest {
+
+	public static class TestableFileWatcherService extends FileWatcherService {
+		public TestableFileWatcherService() {
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java]: unnecessary constructor

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java [hunk 19]
```diff
@@ -0,0 +1,395 @@
...
+			super();
+		}
+		/*
+		 * Getters For testing
+		 */
+		public WatchService getWatchService() {
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java]: missing ```@Override```

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java [hunk 20]
```diff
@@ -0,0 +1,395 @@
...
+			return super.getWatchService();
+		}
+
+		public Thread getServiceThread() {
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java]: missing ```@Override```

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java [hunk 21]
```diff
@@ -0,0 +1,395 @@
...
+			return super.getServiceThread();
+		}
+
+		public HashMap<Path, List<RegistrationRequest>> getRequests() {
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java]: missing ```@Override```

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+			return super.getRequests();
+		}
+
+		public HashMap<Path, WatchKey> getSubscriptions() {
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModelTest.java]: missing ```@Override```

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java [hunk 23]
```diff
@@ -0,0 +1,437 @@
...
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java]: unnecessary import

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java [hunk 24]
```diff
@@ -0,0 +1,437 @@
...
+import java.nio.file.WatchEvent.Kind;
+import java.util.ArrayList;
+import java.util.HashMap;
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java]: unnecessary import

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java [hunk 25]
```diff
@@ -0,0 +1,437 @@
...
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.jboss.tools.rsp.server.filewatcher.FileWatcherService.RegistrationRequest;
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java]: unnecessary import

File: framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java [hunk 26]
```diff
@@ -0,0 +1,437 @@
...
+			WatchEvent<?> event = keyvent.getEvent();
+			Path matched = ((Path)event.context());
+			Path absolute = ((Path)keyvent.key.watchable()).resolve(matched);
+			if( absolute.equals(path) && kind.equals(event.kind()))
+				return true;
+		}
+		return false;
+	}
+	
+	private void verifyModel(TestableFileWatcherService service, Path path,
```
Comment (adietish) [inline framework/tests/org.jboss.tools.rsp.server.test/src/main/java/org/jboss/tools/rsp/server/filewatcher/FilewatcherModificationsTest.java]: extract this method into a ```FileWatcherServiceAssertions``` and reuse it (removing the identical method) in ```FilewatcherModelTest```?

File: framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/filewatcher/IFileWatcherEventListener.java [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+ * All rights reserved. This program is made available under the terms of the
+ * Eclipse Public License v1.0 which accompanies this distribution, and is
+ * available at http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors: Red Hat, Inc.
+ ******************************************************************************/
+package org.jboss.tools.rsp.server.spi.filewatcher;
+
+public interface IFileWatcherEventListener {
+	public void fireEvent(FileWatcherEvent event);
```
[the remark this answers is not in the document]
Comment (robstryker) [reply framework/bundles/org.jboss.tools.rsp.server.spi/src/main/java/org/jboss/tools/rsp/server/spi/filewatcher/IFileWatcherEventListener.java]: will use fileChanged.  +1 on suggestion.

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java (hunk 9 above)
Comment (robstryker) [reply framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: Not sure why you focus on this so much ;)   ArrayList.isEmpty() just returns size == 0   lol.   So it's the same thing.

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java (hunk 13 above)
Comment (robstryker) [reply framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: The thread is no longer in the process of closing. It is now closed.

File: framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java (hunk 16 above)
Comment (robstryker) [reply framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java]: In the case of ENTRY_CREATE, ENTRY_DELETE, and ENTRY_MODIFY events the context is a Path that is the relative path between the directory registered with the watch service, and the entry that is created, deleted, or modified.

All watchkeys we have are created listening to only those three events. So it is guaranteed to be a Path

**Comment by adietish** [comment]
works for me. +1

**Pull request merged**